pkg/kubelet/server/server_test.go

```diff
@@ -855,6 +863,216 @@ func TestContainerLogsWithInvalidTail(t *testing.T) {
+				{stream: v1.LogStreamTypeStderr, msg: "err1\n"},
+				{stream: v1.LogStreamTypeStderr, msg: "err2\n"},
+				{stream: v1.LogStreamTypeStdout, msg: "out1\n"},
+			},
+			query: "?stream=stderr",
+			expectedLogOptions: &v1.PodLogOptions{
+				Stream: &streamStderr,
+			},
+			expectedOutput: "err1\nerr2\n",
+		},
+		{
+			name: "return all logs",
+			logs: []logEntry{
+				{stream: v1.LogStreamTypeStdout, msg: "out1\n"},
+				{stream: v1.LogStreamTypeStderr, msg: "err1\n"},
+				{stream: v1.LogStreamTypeStdout, msg: "out2\n"},
+			},
+			query: "?stream=All",
+			expectedLogOptions: &v1.PodLogOptions{
+				Stream: &streamAll,
+			},
+			expectedOutput: "out1\nerr1\nout2\n",
+		},
+		{
+			name: "stdout logs with legacy tail",
+			logs: []logEntry{
+				{stream: v1.LogStreamTypeStdout, msg: "out1\n"},
+				{stream: v1.LogStreamTypeStderr, msg: "err1\n"},
+				{stream: v1.LogStreamTypeStdout, msg: "out2\n"},
+			},
+			query: "?stream=All&tail=1",
+			expectedLogOptions: &v1.PodLogOptions{
+				Stream:    &streamAll,
+				TailLines: pointer.Int64(1),
+			},
+			expectedOutput: "out2\n",
+		},
+		{
+			name: "last line of logs does not match the stream filter",
+			logs: []logEntry{
+				{stream: v1.LogStreamTypeStdout, msg: "out1\n"},
+				{stream: v1.LogStreamTypeStdout, msg: "out2\n"},
+				{stream: v1.LogStreamTypeStderr, msg: "err1\n"},
+			},
+			query: "?stream=stdout&tailLines=1",
+			expectedLogOptions: &v1.PodLogOptions{
+				Stream:    &streamStdout,
+				TailLines: pointer.Int64(1),
+			},
+			expectedOutput: "",
+		},
+		{
+			name: "last line of logs matches the stream filter",
+			logs: []logEntry{
+				{stream: v1.LogStreamTypeStdout, msg: "out1\n"},
+				{stream: v1.LogStreamTypeStderr, msg: "err1\n"},
+				{stream: v1.LogStreamTypeStdout, msg: "out2\n"},
+			},
+			query: "?stream=stdout&tailLines=1",
+			expectedLogOptions: &v1.PodLogOptions{
+				Stream:    &streamStdout,
+				TailLines: pointer.Int64(1),
+			},
+			expectedOutput: "out2\n",
+		},
+		{
+			name: "return the logs within the last 2 lines that matches the stream filter",
+			logs: []logEntry{
+				{stream: v1.LogStreamTypeStdout, msg: "out1\n"},
+				{stream: v1.LogStreamTypeStderr, msg: "err1\n"},
+				{stream: v1.LogStreamTypeStdout, msg: "out2\n"},
+			},
+			query: "?stream=stderr&tailLines=2",
+			expectedLogOptions: &v1.PodLogOptions{
+				Stream:    &streamStderr,
+				TailLines: pointer.Int64(2),
+			},
+			expectedOutput: "err1\n",
+		},
+		{
+			name: "return the first 6 bytes of the logs within last 3 lines that matches the stream filter",
+			logs: []logEntry{
+				{stream: v1.LogStreamTypeStderr, msg: "err1\n"},
+				{stream: v1.LogStreamTypeStdout, msg: "out1\n"},
+				{stream: v1.LogStreamTypeStderr, msg: "err2\n"},
+				{stream: v1.LogStreamTypeStdout, msg: "out2\n"},
+			},
+			query: "?stream=stdout&tailLines=3&limitBytes=6",
+			expectedLogOptions: &v1.PodLogOptions{
+				Stream:     &streamStdout,
+				TailLines:  pointer.Int64(3),
```

Comment:
so we asked for `3` lines, but only got `1` out of `2`? I think it will be confusing for users

Reply:
The current process when `stream`, `tailLines`, `limitBytes` are all specified is:
1. Get last `tailLines` of logs
2. Filter the above logs by `stream`
3. Send the first `limitBytes` of filtered logs to user

As I noted in the PR description:
> Note that if Stream is used in conjunction with TailLines, the output is the logs from the last TailLines that match the given Stream, rather than the last TailLines logs of the given Stream.
>
> The implementation of the latter requires parsing log entry to extract the log stream and a buffering mechanism (the memory required is proportional to TailLines) when reading container logs, which is not efficient and prone to DoS attack.

Given that every line of log is either encoded in json-file or cri format, we need to parse every line of logs to examine its `stream`, which is a heavy operation. What's worse, to get the last N lines of a given `stream`, we ultimately need to buffer matched logs, which might be memory-intensive and susceptible to DoS attack. I cannot think of an efficient solution to get the last N lines of a given `stream` by now.

Reply:
From scenarios perspective, would it be confusing to get `0` errors if you are looking for the "last error" using `TailLines`? What would be the suggested use of `TailLines` in the proposed model and when it will ever be used alongside the stream filtering? Would it be more logical to disallow to specify both if it is prohibitively expensive to implement?

if you think we can keep this implementation, how should we change the description of `TailLines` field? Can you please write down the proposal here so we can discuss the tradeoffs.

Reply:
Okay I will file a PR to update the KEP soon. Let's discuss the implementation details and trade-off there.

Reply:
> Would it be more logical to disallow to specify both if it is prohibitively expensive to implement?

@SergeyKanzhelev Hi I decided to take your advice, and updated the KEP https://github.com/kubernetes/enhancements/pull/4879

Reply:
cool. Also please add validation that if both are set we will return an error